src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs

```diff
@@ -226,6 +226,10 @@ public class OutputCacheFilter : FilterProvider, IActionFilter, IResultFilter, I
                         // To prevent access to the original lifetime scope a new work context scope should be created here and dependencies
                         // should be resolved from it.
 
+                        // Recheck the response status code incase it was modified before the callback.
+                        if (response.StatusCode != 200)
+                            return;
```

Comment:
Let's also add a debug log message here I think. Something like this:
"Response for item '{_cacheKey}' will not be cached because status code was set to {response.StatusCode} during rendering."